Catnip docs (#353)
configcat/docs: website/docs/targeting/feature-flag-evaluation.md

---
id: feature-flag-evaluation
title: Feature Flag Evaluation
description: This document offers an in-depth explanation of how the ConfigCat SDK determines the value of a feature flag.
---

# Feature Flag Evaluation

This document offers an in-depth explanation of how the SDK determines the value of a feature flag when executing the `GetValue` function. Understanding this process requires prior knowledge of targeting concepts.

> The feature flag's value is influenced by:

Reviewer: ```suggestion
The feature flag's value is determined by:
```

- The feature flag's rules defined on the Dashboard,
- The [User Object] provided to the `GetValue` function, and
- The *default value* passed to the `GetValue` function.

The feature flag's value always comes from exactly one rule, following this algorithm:

1. **Evaluation of Targeting Rules:** If targeting rules are present, the SDK evaluates them one by one, from top to bottom. It checks if all the conditions in the rule's IF part are met (i.e. all the conditions evaluate to true).

>    - If the conditions are met, the THEN part determines the value to return. Note: If the THEN part contains [Percentage Options] and the necessary user attribute is missing, the SDK will skip the targeting rule and continue with the next rule - even though the targeting rule's conditions are met!

Reviewer: ```suggestion
   - If the conditions are met, the THEN part determines the value to return. Note: If the THEN part contains percentage options but the [percentage evaluation attribute] is missing, the SDK will skip the targeting rule and continue with the next rule - even though the targeting rule's conditions are met!
```

   - If the conditions aren't met, the SDK moves to the next targeting rule, or to step 2 (below) if there are no more targeting rules.

> 2. **Evaluation of Percentage Options:** If a percentage option rule exists, the SDK executes the [Evaluation of percentage options] algorithm to determine which percentage option applies to the user and returns the value associated with that option. If the necessary user attribute is missing, the SDK skips to step 3 (below).

Reviewer: ```suggestion
2. **Evaluation of Percentage Options:** If a percentage options rule exists, the SDK executes the [Evaluation of percentage options] algorithm to determine which percentage option applies to the user and returns the value associated with that option. If the necessary user attribute is missing, the SDK skips to step 3 (below).
```

3. **Returning simple value:** At this stage, the only remaining "rule" is the simple value specified at the end of the feature flag, which the SDK then returns.

In the event of an unexpected error during evaluation, the SDK returns the default value passed to the `GetValue` function.

## Evaluation of a Targeting Rule

The SDK evaluates the conditions in the rule's IF part one by one, from top to bottom. The result of a condition can be one of the following: `true`, `false` or `cannot evaluate`.

The result `cannot evaluate` occurs when the necessary user attribute is missing, invalid or incorrectly formatted (the SDK logs these issues as warnings).

A targeting rule matches only when all its conditions evaluates to `true`. In any other cases, it doesn't match.

### Evaluation of a User Condition

> The SDK looks up the comparison attribute (the user attribute referenced by the condition) in the [User Object]. It compares it to the comparison value that is set on the Dashboard. The comparison is done according to the selected comparator, resulting in a `true` or `false` value. This will be the result of the condition.

Reviewer: ```suggestion
The SDK looks up the comparison attribute (the user attribute referenced by the condition) in the [User Object]. It compares the attribute value to the comparison value that is set on the Dashboard. The comparison is done according to the selected comparator, resulting in a `true` or `false` value. This will be the result of the condition.
```

The result of the condition will be `cannot evaluate` in case the comparison attribute is missing (`null`, `undefined`, `""`) or invalid (not of the type expected by the comparator or not formatted properly). In such cases, the targeting rule containing the condition will be skipped, and the evaluation will continue with the next rule.

### Evaluation of a Flag Condition

> Using the same User Object used to evaluate the dependent flag, the SDK evaluates the prerequisite flag (the feature flag referenced by the condition). Then, the result is sent to the comparison value set on the Dashboard. The comparison is done according to the selected comparator, resulting in a `true` or `false` value. This will be the result of the condition.

Reviewer: ```suggestion
Using the same User Object used to evaluate the dependent flag, the SDK evaluates the prerequisite flag (the feature flag referenced by the condition). Then, the result is checked against the comparator. In case the prerequisite flag is not a boolean setting, the result is compared to the comparison value that is set on the Dashboard. The comparison results in a `true` or `false` value. This will be the result of the condition.
```

> In case the referenced feature flag is missing or there is a type mismatch between the return value and the comparison value, the evaluation process stops, and the SDK will return the [default value]. (Though this can happen only when using the [flag overrides] feature with invalid data.)

Reviewer: ```suggestion
In case the prerequisite flag is missing or there is a type mismatch between the return value and the comparison value, the evaluation process stops, and the SDK will return the [default value]. (Though this can happen only when using the [flag overrides] feature with invalid data.)
```

### Evaluation of a Segment Condition

> The SDK looks up the segment referenced by the condition and evaluates the condition described by the segment similarly to [user conditions]. In the case of `IS IN SEGMENT` the result of the segment condition will be the same as the result of the segment. The result will be negated in the case of `IS NOT IN SEGMENT`.

Reviewer: ```suggestion
The SDK looks up the segment referenced by the condition and evaluates the condition described by the segment similarly to [user conditions]. Then, the result is checked against the comparator. In the case of `IS IN SEGMENT` the result of the segment condition will be the same as the result of the segment. The result will be negated in the case of `IS NOT IN SEGMENT`.
```

If the segment evaluates to `cannot evaluate`, so is the segment condition.

The evaluation process stops if the referenced segment is missing, and the SDK will return the [default value]. (Though this can happen only when using the [flag overrides] feature with invalid data.)

## Evaluation of Percentage Options

Percentage options are designed to be consistent and sticky across all SDKs, which means that users with the same attributes fall in the same group and get the same feature flag value across the supported platforms.

> The SDK looks up the [percentage evaluation attribute] in the [User Object] and:
> - The SDK creates a hash from the combination of the value of the percentage evaluation attribute and the specific feature flag's key.

Reviewer: ```suggestion
The SDK looks up the [percentage evaluation attribute] in the [User Object], then:
- The SDK creates a hash from the combination of the percentage evaluation attribute's value and the specific feature flag's key.
```

---
id: feature-flag-evaluation

> title: Setting Evaluation
> description: This document offers an in-depth guide on how the SDK determines the value of a setting.
> ---
> 
> # Setting Evaluation

Reviewer: ```suggestion
title: Feature Flag Evaluation
description: This document offers an in-depth guide on how the SDK determines the value of a setting.
---

# Feature Flag Evaluation
```
